index/codec/doc.go

```diff
@@ -0,0 +1,304 @@
+		return errEmptyDocument
+	}
+	defer w.enc.reset()
+
+	w.enc.putUvarint(uint64(len(d.Fields)))
+
+	for _, f := range d.Fields {
+		w.enc.putBytes(f.Name)
+		w.enc.putBytes(f.Value)
+	}
+
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	w.docs++
+	return nil
+}
+
+func (w *docWriter) Close() error {
+	w.enc.putUint32(w.checksum.get())
+	if err := w.write(); err != nil {
+		return err
+	}
+	w.checksum.reset()
+
+	w.enc.putUint64(w.docs)
+	w.enc.putUint32(w.checksum.get())
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *docWriter) write() error {
+	b := w.enc.bytes()
+	n, err := w.writer.Write(b)
+	if err != nil {
+		return err
+	}
+	if n < len(b) {
+		return io.ErrShortWrite
+	}
+	w.checksum.update(b)
+	return nil
+}
+
+type docReader struct {
+	version uint32
+
+	data []byte
+	dec  *decoder
+
+	curr     uint64
+	total    uint64
+	checksum *checksum
+}
+
+func newDocReader(data []byte) *docReader {
+	return &docReader{
+		data: data,
+		dec:  new(decoder),
+	}
+}
+
+func (r *docReader) Init() error {
+	if len(r.data) < docsHeaderSize+docsTrailerSize {
+		return io.ErrShortBuffer
+	}
+
+	if err := r.decodeHeader(); err != nil {
+		return err
+	}
+	if err := r.decodeTrailer(); err != nil {
+		return err
+	}
+	if err := r.verifyPayload(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *docReader) decodeHeader() error {
+	r.dec.reset(r.data[docsHeaderSize:])
+
+	// Verify magic number.
+	n, err := r.dec.uint32()
+	if err != nil {
+		return err
+	}
+	if n != docsMagicNumber {
+		return errInvalidMagicNumber
+	}
+
+	// Verify file format version.
+	v, err := r.dec.uint32()
+	if err != nil {
+		return err
+	}
+	if v != docsFileFormatVersion {
+		return fmt.Errorf("version of file format: %v does not match expected version: %v", v, docsFileFormatVersion)
+	}
+	r.version = v
+
+	r.checksum.update(r.data[:8])
+	if err := r.verifyChecksum(r.dec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *docReader) decodeTrailer() error {
+	r.dec.reset(r.data[len(r.data)-docsTrailerSize:])
+
+	// Get number of documents.
+	n, err := r.dec.uint64()
+	if err != nil {
+		return err
+	}
+	r.total = n
+
+	r.checksum.update(r.data[:8])
+	if err := r.verifyChecksum(r.dec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *docReader) verifyPayload() error {
+	start := docsHeaderSize
+	end := len(r.data) - docsTrailerSize - 4
+	payload := r.data[start:end]
+	r.checksum.update(payload)
+
+	checksum := r.data[end : end+4]
+	r.dec.reset(checksum)
+	if err := r.verifyChecksum(r.dec); err != nil {
+		return err
+	}
+
+	// Update decoder for use in subsequent calls to Read.
+	r.dec.reset(payload)
+	return nil
+}
+
+func (r *docReader) verifyChecksum(dec *decoder) error {
+	c, err := dec.uint32()
+	if err != nil {
+		return err
+	}
+	if c != r.checksum.get() {
+		return errChecksumMismatch
+	}
+	r.checksum.reset()
+	return nil
+}
+
+func (r *docReader) Read() (doc.Document, error) {
```

Comment:
maybe using this would look better as an iterator?

Reply:
talked offline, more consistent with code elsewhere (m3db/et all) to use current impl